packages/peregrine/lib/talons/Gallery/useGalleryItem.js

```diff
@@ -0,0 +1,71 @@
+import { useCallback, useEffect, useMemo, useRef, useState } from 'react';
+import { useLazyQuery } from '@apollo/react-hooks';
+
+/**
+ *
+ * @param {Object} props
+ * @param {Object} props.item - the product item being rendered
+ * @param {Object} props.queries - the queries for the gallery item
+ *
+ * @returns {{
+ *  ref: the ref for the gallery item, used by the intersection observer
+ * }}
+ */
+export const useGalleryItem = props => {
+    const { item, queries } = props;
+
+    const { prefetchProductQuery } = queries;
+    const { url_key } = item;
+
+    // Prefetch the underlying product for this gallery item, but only if we
+    // don't have any product data yet, and only once the gallery item is in
+    // view.
+    const [runPrefetchQuery] = useLazyQuery(prefetchProductQuery);
+    const [hasPrefetched, setHasPrefetched] = useState(false);
+    const galleryItemRef = useRef();
+
+    const handleScrollIntoView = useCallback(
+        entries => {
+            entries.forEach(entry => {
+                if (entry.isIntersecting && !hasPrefetched) {
+                    setHasPrefetched(true);
+                    runPrefetchQuery({
+                        variables: {
+                            urlKey: url_key,
+                            onServer: false
+                        },
+                        // If we get a cache-hit, the benefit of the prefetch
+                        // is achieved already, so no need to hit network.
+                        fetchPolicy: 'cache-first'
```

Comment:
As @tjwiebell mentioned, there is possibility to manage this with edge caching to lighten the server load, but that may not be possible for everyone.

I can add a `shouldPrefetch` boolean, check "time-in-viewport", or try other optimizations, but at the end of the day, from a cold cache, this code still has the possibility to emit a large number of queries at once. 

I would love to talk about options outside of reverting the PR, as I believe this code is more explicit and provides a better developer experience.

Reply:
I'm curious what people think about prefetching on a more specific user interaction. Such as hover/focus/click/tab on a given CTA. 

Pros:
* Still serves the purpose of preloading, but is more specific and should cause less strain on the application server
* For requests that are a cache hit, this shouldn't be too much of a difference than the current approach

Cons:
* For requests that are a cache miss, this would cause a more noticeable lag for the user, when compared with the current approach.
* Won't work as well for mobile users

Reply:
Now that you mention it, I think using on hover/focus might be a good optimization here. I'm also questioning whether mobile prefetch is even necessary. I like to be data conscious for mobile users and disabling prefetch for mobile (since there is no hover/focus) might just be fine.